### PR TITLE
User searching

### PR DESCRIPTION
**Pull request by rushsteve1**
This adds a search bar to the users docker that allows you to search/filter users in the room easily.
Works quite well and is very fast even though it's just linear (tested in #matrix:matrix.org).

Also I minor QoL of setting the device name to the machine's hostname when there are no accounts to pull that info from.

### CODE REVIEW

**Comment by KitsuneRal** [comment]
First off, let's put the login dialog change to a separate PR - I'm not really happy to pick the hostname by default.

**Comment by rushsteve1** [comment]
Moved the device name commit to PR #438 as requested

File: client/models/userlistmodel.cpp [hunk 1]
```diff
@@ -144,6 +140,23 @@ void UserListModel::userRemoved(QMatrixClient::User* user)
...
+    QList<User*> filteredUsers;
+    for(auto i = m_users.begin(); i != m_users.end(); i++) {
+        if ((*i)->fullName(m_currentRoom).contains(filterString) ||
+            (*i)->displayname(m_currentRoom).contains(filterString)
+        ) {
+            filteredUsers.append(*i);
+        }
+    }
+    m_users = sortUsers(filteredUsers);
+    endResetModel();
```
Comment (KitsuneRal) [inline client/models/userlistmodel.cpp]: `endResetModel()` without `beginResetModel()` in the same block looks very unnerving. Either you have already forgotten to put `beginResetModel()` in a proper place, or you are very much up to.

File: client/models/userlistmodel.cpp [hunk 2]
```diff
@@ -170,3 +183,14 @@ int UserListModel::findUserPos(const QString& username) const
...
+    if ( m_currentRoom == nullptr) {
+        qWarning() << "Cannot sort users without m_currentRoom set";
+        return users;
+    }
+    QElapsedTimer et; et.start();
+    std::sort(users.begin(), users.end(), m_currentRoom->memberSorter());
+    qDebug() << "Sorting" << users.size() << "user(s) in"
+                << m_currentRoom->displayName() << "took" << et;
+    return users;
+}
```
Comment (KitsuneRal) [inline client/models/userlistmodel.cpp]: Assuming `filter()` is rewritten in the way above, you won't need a separate `sortUsers()` method - `filter("")` will do the same with no overheads. Taking `QElapsedTimer` and the debug log line from here to `filter()` would be very nice, of course.

File: client/userlistdock.cpp [hunk 3]
```diff
@@ -52,6 +66,8 @@ UserListDock::UserListDock(QWidget* parent)
              this, &UserListDock::refreshTitle );
     connect( m_model, &QAbstractListModel::modelReset,
              this, &UserListDock::refreshTitle );
+    connect(filterline, &QLineEdit::textEdited,
+             m_model, &UserListModel::filter);
```
[the remark this answers is not in the document]
Comment (rushsteve1) [reply client/userlistdock.cpp]: A fair number of users have single-character usernames. Especially if their name is an emoji (which is apparently possible).

File: client/models/userlistmodel.cpp [hunk 4]
```diff
@@ -144,6 +140,23 @@ void UserListModel::userRemoved(QMatrixClient::User* user)
...
+        m_users = sortUsers(m_currentRoom->users());
+    }
+    QList<User*> filteredUsers;
+    for(auto i = m_users.begin(); i != m_users.end(); i++) {
+        if ((*i)->fullName(m_currentRoom).contains(filterString) ||
+            (*i)->displayname(m_currentRoom).contains(filterString)
+        ) {
+            filteredUsers.append(*i);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (rushsteve1) [reply client/models/userlistmodel.cpp]: Thank you for this snippet it was very helpful.

**Comment by rushsteve1** [comment]
This PR is blocking PR #442

**Pull request merged**